src/Init/Init.c

```diff
@@ -0,0 +1,299 @@
+#include "SystemConfig/SystemConfig.h"
+#include "Pio/Pio.h"
+#include "Nvic/Nvic.h"
+#include "Uart/Uart.h"
+#include "Sdramc/Sdramc.h"
+
+#include <assert.h>
+#include <string.h>
+
+extern Uart consoleUart;
+
+static void Init_setup_watchdog(void);
+static void Init_setup_pmc(void);
+static void Init_setup_fpu(void);
+static void Init_setup_pio(void);
+static void Init_setup_nvic(void);
+static void Init_setup_uart(void);
+static void Init_setup_sdram(void);
+
+static void Init_setup_sdram_pio(void);
+static void Init_setup_sdram_pio_portA(void);
+static void Init_setup_sdram_pio_portC(void);
+static void Init_setup_sdram_pio_portD(void);
+static void Init_setup_sdram_pio_portE(void);
+static void Init_setup_sdram_pio_pin(Pio* const pio, const uint32_t pin, const Pio_Control control);
+static void Init_setup_sdram_pmc(void);
+static void Init_setup_sdram_config(Sdramc* const sdramc);
+
+#define UART_BAUDRATE 38400
+
+void
+Init_setup_hardware(void)
+{
+    Init_setup_watchdog();
+    Init_setup_pmc();
+    Init_setup_fpu();
+    Init_setup_pio();
+    Init_setup_nvic();
+    Init_setup_uart();
+    Init_setup_sdram();
+}
+
+static void
+Init_setup_watchdog(void)
+{
+    const Wdt_Config wdtConfig = {
+        .counterValue = 0x0FFF,
+        .deltaValue = 0x0FFF,
+        .isResetEnabled = false,
+        .isFaultInterruptEnabled = false,
+        .isDisabled = true,
+        .isHaltedOnIdle = false,
+        .isHaltedOnDebug = false,
+    };
+
+    Wdt wdt;
+    Wdt_init(&wdt);
+    Wdt_setConfig(&wdt, &wdtConfig);
+}
+
+static void
+Init_setup_fpu(void)
+{
+    Fpu fpu;
+    Fpu_init(&fpu);
+    Fpu_startup(&fpu);
+}
+
+static void
+Init_setup_pmc(void)
+{
+    int errCode = 0;
+
+    Pmc_Config pmcConfig = SystemConfig_getPmcDefaultConfig();
+    Pmc_setConfig(&pmcConfig, &errCode);
+
+    assert(errCode == 0);
+}
+
+static void
+Init_setup_pio(void)
+{
+    Pmc_enablePeripheralClk(Pmc_PeripheralId_PioD);
+    Pmc_enablePeripheralClk(Pmc_PeripheralId_Uart4);
+
+    Pio pio;
+    Pio_init(Pio_Port_D, &pio);
+
+    Pio_Port_Config pioConfig = {
+        .pinsConfig = {
+            .pull = Pio_Pull_Up,
+            .filter = Pio_Filter_None,
+            .isMultiDriveEnabled = false,
+            .isSchmittTriggerDisabled = false,
+            .irq = Pio_Irq_None,
+        },
+        .debounceFilterDiv = 0
+    };
+
+    pioConfig.pins = PIO_PIN_18;
+    pioConfig.pinsConfig.control = Pio_Control_PeripheralC;
+    pioConfig.pinsConfig.direction = Pio_Direction_Input;
+    Pio_setPortConfig(&pio, &pioConfig);
+
+    pioConfig.pins = PIO_PIN_19;
+    pioConfig.pinsConfig.control = Pio_Control_PeripheralC;
+    pioConfig.pinsConfig.direction = Pio_Direction_Output;
+    Pio_setPortConfig(&pio, &pioConfig);
+}
+
+static void
+Init_setup_nvic(void)
+{
+    Nvic_enableInterrupt(Nvic_Irq_Uart4);
+    Nvic_setInterruptPriority(Nvic_Irq_Uart4, 0);
+}
+
+static void
+Init_setup_uart(void)
+{
+    Uart_Config uartConfig = { .isTxEnabled = true,
```

Comment:
I understand that this is for testing... but I would prefer to put here a TODO indicating that UART should be removed from here. All UART initialization should be done by HAL, which may additionally initialize USART for debug output.

Reply:
Sure, I wanted to change that after implementing HAL. I will add TODO.